store/tikv/latch/latch.go

```diff
@@ -0,0 +1,186 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package latch
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"sync"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spaolacci/murmur3"
+)
+
+// Latch stores a key's waiting transactions
+type Latch struct {
+	waiting      []uint64
+	lastCommitTs uint64
+	lock         sync.Mutex
+}
+
+// acquire tries to get current key's lock for the transaction with startTs
+// acquire is true when success
+// timeout is true when the startTs is already timeout
+// newWait is true when current transaction is new for the current latch
+func (l *Latch) acquire(startTs uint64) (acquire, timeout, newWait bool) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	timeout = startTs <= l.lastCommitTs
+	if timeout {
+		return
+	}
+	if len(l.waiting) == 0 || l.waiting[0] != startTs {
+		l.waiting = append(l.waiting, startTs)
+		newWait = true
+	}
+
+	acquire = l.waiting[0] == startTs
+	return
+}
+
+// release releases the transaction with startTs and commitTs from current latch.
+// isEmpty is true when the waiting queue is empty after release current transaction,
+// otherwise return the front transaction in queue.
+func (l *Latch) release(startTs uint64, commitTs uint64) (isEmpty bool, front uint64) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	if startTs != l.waiting[0] {
+		panic(fmt.Sprintf("invalid front ts %v, latch:%+v", startTs, l))
+	}
+	// for rollback or timeout, the commitTs maybe zero
+	if commitTs > l.lastCommitTs {
+		l.lastCommitTs = commitTs
+	}
+	l.waiting = l.waiting[1:]
+	if len(l.waiting) == 0 {
+		// Should we reset the lastCommitTs since next key
+		// may different from current key?
+		// l.lastCommitTs = 0
+		isEmpty = true
+	} else {
+		front = l.waiting[0]
+		isEmpty = false
+	}
+	return
+}
+
+// try append startTs into latch, return false
+// when startTs is already timeout
+func (l *Latch) append(startTs uint64) bool {
+	if startTs <= l.lastCommitTs {
+		return false
+	}
+	l.waiting = append(l.waiting, startTs)
+	return true
+}
+
+// Lock is the locks' information required for a transaction
+type Lock struct {
+	// the slot IDs of the latches(keys) that a startTs must acquire before being able to processed
+	requiredSlots []int
+	/// The number of latches that the transaction has acquired.
+	acquiredCount int
+	/// The number of latches that the waiting queue contains current transaction
+	waitedCount int
+	// Current transaction's startTs
+	startTs uint64
+}
+
+// NewLock creates a new lock
+func NewLock(startTs uint64, requiredSlots []int) Lock {
+	return Lock{
+		requiredSlots: requiredSlots,
+		acquiredCount: 0,
+		waitedCount:   0,
+		startTs:       startTs,
+	}
+}
+
+// Latches which are used for concurrency control
+// Each latch is indexed by a slit ID, hence the term latch and slot are used in interchangeable,
+// but conceptually a latch is a queue, and a slot is an index to the queue
+type Latches struct {
```

Comment:
It is better to use type alias here. `type Latches []Latch`. The size of `Laches` can be got from `Len(l)` if `l`is a instance of `Latches`.